VersionControlUI/src/DiffManager.cpp

```diff
@@ -0,0 +1,300 @@
+		};
+
+	// load newer version
+	Model::TreeManager* newVersionManager = new Model::TreeManager{};
+	newVersionManager->load(newFileStore, _project, false);
+
+	// load older version
+	Model::TreeManager* oldVersionManager = new Model::TreeManager{};
+	oldVersionManager->load(oldFileStore, _project, false);
+
+	Model::Reference::resolvePending();
+
+	FilePersistence::Diff diff = repository.diff(_oldVersion, _newVersion);
+
+	auto changes = diff.changes();
+
+	const int INDEX_OF_DELETE = 0;
+	const int INDEX_OF_ADD = 1;
+	const int INDEX_OF_MOVE = 2;
+	const int INDEX_OF_MODIFY = 3;
+
+	QList<QSet<Model::NodeIdType>> nodeLists;
+	for (int i = 0; i < 4; i++)
+		nodeLists.append(QSet<Model::NodeIdType>{});
+
+	QSet<Model::NodeIdType> changedNodesToVisualize;
+
+	for (auto change : changes.values())
+	{
+		if (change->isFake() || change->onlyStructureChange()) continue;
+
+		auto id = change->nodeId();
+
+		switch (change->type())
+		{
+			case FilePersistence::ChangeType::Deletion:
+				nodeLists[INDEX_OF_DELETE].insert(id);
+				break;
+			case FilePersistence::ChangeType::Insertion:
+				nodeLists[INDEX_OF_ADD].insert(id);
+				break;
+			case FilePersistence::ChangeType::Move:
+				nodeLists[INDEX_OF_MOVE].insert(id);
+				break;
+			case FilePersistence::ChangeType::Stationary:
+				nodeLists[INDEX_OF_MODIFY].insert(id);
+				break;
+			case FilePersistence::ChangeType::Unclassified:
+				qDebug() << "Unclassified";
+				break;
+		}
+		Model::NodeIdType nodeId;
+		if (change->type() != FilePersistence::ChangeType::Deletion)
+			if (findChangedNode(newVersionManager, id, nodeId))
+				changedNodesToVisualize.insert(nodeId);
+		if (change->type() != FilePersistence::ChangeType::Insertion)
+			if (findChangedNode(oldVersionManager, id, nodeId))
+				changedNodesToVisualize.insert(nodeId);
+	}
+	//
+	for (int i = 0; i < nodeLists.size(); i++)
+		removeDirectChildrenOfNodesInContainer(&nodeLists[i]);
+
+	Visualization::VisualizationManager::instance().mainScene()->listenToTreeManager(newVersionManager);
+	Visualization::VisualizationManager::instance().mainScene()->listenToTreeManager(oldVersionManager);
+
+	diffViewItem->setMajorAxis(Visualization::GridLayouter::NoMajor);
+
+	// add nodes to visualization, if not avaible show text
+	for (auto id : changedNodesToVisualize)
+	{
+		auto oldNode = const_cast<Model::Node*>(oldVersionManager->nodeIdMap().node(id));
+		auto newNode = const_cast<Model::Node*>(newVersionManager->nodeIdMap().node(id));
+
+		// old version in left column
+		if (!oldNode)
+			oldNode = new Model::Text{"node in old version not found"};
+		diffViewItem->insertNode(oldNode, 0);
+
+		// new version in right column
+		if (!newNode)
+			newNode = new Model::Text{"node in new version not found"};
+
+		diffViewItem->insertNode(newNode, 1);
+	}
+
+	//TODO method for visualizations
+
+	// create highlights and arrows for moved nodes
+	for (auto id : nodeLists[INDEX_OF_MOVE])
```

Comment:
These four loops are very similar to each other, please abstract them into a method/lambda to remove code duplication.